### ISSUE TITLE
(taken from jsch-users): Possible deadlock in KnownHosts

### ISSUE THREAD

**Issue by kimmerin** [mwiede/jsch#131]
Hi,

I've remembered that there was an issue reported on jsch-users where I've answered with a patch. I don't have the original mail in my archive on my current computer but the internet [doesn't forget](https://jsch-users.narkive.com/XfjBWvoZ/deadlock-in-knownhosts).

I've checked the current source and it seems that a part of my fix got included into the source base but not all, especially the removal of the synchronization that lead to the deadlock in the mentioned report above.

Also KnownHost contains a lot of `System.out` and `System.err.println` calls in case of errors that should be changed to logger.log-calls in my eyes.

**Comment by norrisjeremy** [comment]
I'm not sure I see how this deadlock could still be occurring, at least as described in the original JSch-users thread: the `KnownHosts` class no longer contains any synchronized methods.

Are you still able to reproduce this problem with current release?

**Comment by kimmerin** [comment]
> the `KnownHosts` class no longer contains any synchronized methods.

Um...

https://github.com/mwiede/jsch/blob/7e05876c65502cd0f350a5fcc75dc80e6a1da353/src/main/java/com/jcraft/jsch/KnownHosts.java#L551-L558

https://github.com/mwiede/jsch/blob/7e05876c65502cd0f350a5fcc75dc80e6a1da353/src/main/java/com/jcraft/jsch/KnownHosts.java#L582-L588

That's what I see in trunk right at this moment.

> Are you still able to reproduce this problem with current release?

The problem was reported by somebody else and I just provided a solution by looking very closely at the deadlock-message in that mail and its stacktraces. We've had occasional deadlocks as well "out there" and after this fix never heard of one again (but I might remember wrong and this was fixed by me at some other place back then).

As mentioned before in another issue I'm currently doing a lot of customization of the original sources when a new release is available (a lot of effort I try to reduce by contributing the useful stuff to this project - the logging issue being one of them). So if I do testing with SSH I always use a version without above synchronized blocks.

Above synchronized blocks are completely unnecessary. `hash()` get's only called during instantiation so that you don't have the situation that `hash()` and `isMatched` get called in parallel. The blocks are a relic of the time when there was a VM-wide hash-instance.

**Comment by norrisjeremy** [comment]
Neither of those are synchronizing on the `KnownHosts` objects, they are synchronizing on the `HmacSHA1` object.

The original deadlock as you linked from JSch-Users archive is caused by a deadlock between `KnownHost` object and `KnownHosts.pool` object, that was caused because:

Thread 1: locks the `KnownHosts` object (`hkr`) in `Session.checkHost()`, then attempts to lock the `pool` object whilst calling `KnownHosts.check()`.

Thread 2: locks the `pool` object in `KnownHosts.getHostKey()`, then attempts to to lock the `KnownHosts` object because `KnownHosts. getHMACSHA1()` used to be a synchronized method.

So you end up with Thread 1 & Thread 2 waiting on for locks owned by each other.

However, current version of `KnownHosts` in this fork, no longer has `KnownHosts. getHMACSHA1()` as a synchronized method.

Therefore, if you look at how the deadlock originally occurred in that old release of JSch, you will see it can no longer happen, because the call by Thread 2 to `KnownHosts. getHMACSHA1()` will no longer attempt to lock the `KnownHosts` object:

Thread 1: locks the `KnownHosts` object (`hkr`) in `Session.checkHost()`, then attempts to lock the `pool` object whilst calling `KnownHosts.check()`.

Thread 2: locks the `pool` object in `KnownHosts.getHostKey()`, then completes its call to `KnownHosts.getHMACSHA1()` successfully and will eventually drop the lock on the `pool` object when `KnownHosts.getHostKey()` returns to the caller, at which time Thread 1 will gain the lock on the `pool` object and complete it's call to `KnownHosts.check()`.

**Comment by norrisjeremy** [comment]
Btw, I'm pretty sure this was noted as being fixed in the old 0.1.54 release as documented here:
https://github.com/mwiede/jsch/blob/7e05876c65502cd0f350a5fcc75dc80e6a1da353/ChangeLog#L31

**Comment by norrisjeremy** [comment]
Btw, I do agree that cleaning up the calls to `System.out.println()` & `System.err.println()` in the `KnownHosts` class is an excellent idea.

Also, it probably would be an improvement to stop calling `getHMACSHA1()` from the `isMatched()` & `hash()` methods, and simply use the `hmacsha1` property of the object itself.

In fact, you can see that it's even possible to cause NPE's, because `getHMACSHA1()` can return `null`, but the calls to it in `isMatched()` & `hash()` both directly use the return value without confirming it isn't `null` first.

**Comment by kimmerin** [comment]
As said, Here[TM] I'm testing with a customized version of JSch that contain the complete patch as shared on the mail list back then but I assume that you're right with your analysis that having an hash-instance per HashedHostKey-instance is sufficient to prevent occurences.

I'll take your other answer as an invitation for implementation ;-) so I'll have a look into it.

**Comment by norrisjeremy** [comment]
I'm confused, why are you using `HMACSHA256` for host key hashing?

**Comment by norrisjeremy** [comment]
I think you've misunderstood the purpose of `JSch.setConfig("hmac-sha1", ...)`.
It's not to allow a user to pick a different HMAC algorithm, like HMAC-SHA256,: it's to let a user direct JSch to use a customized implementation of HMAC-SHA1, like with a different crypto provider.

There's no need to make any changes here: if a user incorrectly execute `JSch.setConfig("hmac-sha1", HMACSHA256.class.getName())`, lots of things will break, not just this.

**Comment by kimmerin** [comment]
Misunderstanding is the wrong word. I've looked at the source, checked what's possible and wrote tests around that.

As I said: If the use of non-SHA-1-hashes shouldn't be allowed, trying to set one should lead to an exception. Otherwise - looking at KnownHosts - you will corrupt your known-hosts-file e.g. as soon as you do a remove-call, that leads to a sync where the wrong hashes are written to the file.

**Comment by norrisjeremy** [comment]
Again, there is no need to do anything here: if you incorrectly execute `JSch.setConfig("hmac-sha1", ...)` to point at a `MAC` implementation that is not actually implementing HMAC-SHA1, you've broken alot more than the known_hosts file.
For example, MAC negotiation of `hmac-sha1`  would utterly break and your SSH session would not function.

Also, I'm not really sure how you would propose to to implement any sort of runtime detection of what a user configures `JSch.getConfig("hmac-sha1')` as JSch uses reflection to instantiate the class and it can be pointed at anything.

I.e.  it's impossible to check that when a user executes `JSch.setConfig("hmac-sha1", "my.own.HMACSHA1")`, to determine that `my.own.HMACSHA1` is a `com.jcraft.jsch.MAC` implementation that conforms to HMAC-SHA1.

**Comment by kimmerin** [comment]
> a MAC implementation that is not actually implementing HMAC-SHA1, you've broken alot more than the known_hosts file.
For example, MAC negotiation of hmac-sha1 would utterly break and your SSH session would not function.

Personally I see a corrupt hosts-file much more serious than a handshake-failure. After fixing the MAC-configuration the latter starts working again while the former will still leave you with a corrupt hosts-file.

**Comment by norrisjeremy** [comment]
> > a MAC implementation that is not actually implementing HMAC-SHA1, you've broken alot more than the known_hosts file.
> > For example, MAC negotiation of hmac-sha1 would utterly break and your SSH session would not function.
> 
> Personally I see a corrupt hosts-file much more serious than a handshake-failure. After fixing the MAC-configuration the latter starts working again while the former will still leave you with a corrupt hosts-file.
> 
> Edit: can you show me where `JSch.getConfig("hmac-sha1")` is used other than in KnownHosts? I've only found it to be used in KnownHosts and `KnownHosts.getHMACSHA1` is only called in the context of KnownHosts as well. So how can setting `hmac-sha1` to `HMACSHA256` have any effect on the MAC negotiation you've mentioned? I've searched for the text occurrence of `getConfig("hmac-sha1")` to also get `Session.getConfig`-calls.

https://github.com/mwiede/jsch/blob/master/src/main/java/com/jcraft/jsch/Session.java#L1497
https://github.com/mwiede/jsch/blob/master/src/main/java/com/jcraft/jsch/Session.java#L1526
https://github.com/mwiede/jsch/blob/master/src/main/java/com/jcraft/jsch/Session.java#L2905

**Comment by kimmerin** [comment]
> https://github.com/mwiede/jsch/blob/master/src/main/java/com/jcraft/jsch/Session.java#L1497

That explains why I haven't found anything ;-)

I'll remove the test with HMACSHA256 then but I still like to add a "is-this-SHA-1" check to prevent the corrupt-known-hosts-file-situation from happening. Is that OK?

**Comment by norrisjeremy** [comment]
> > https://github.com/mwiede/jsch/blob/master/src/main/java/com/jcraft/jsch/Session.java#L1497
> 
> That explains why I haven't found anything ;-)
> 
> I'll remove the test with HMACSHA256 then but I still like to add a "is-this-SHA-1" check to prevent the corrupt-known-hosts-file-situation from happening. Is that OK?

I would rather not, as it would simply degrade performance. If a user incorrectly sets `hmac-sha1` to point at something that isn't implementing HMAC-SHA1, then too bad.

**Comment by kimmerin** [comment]
> I would rather not, as it would simply degrade performance.

Loading the known-hosts-file isn't part of a performance-critical operation. The read-operation of the file should take magnitudes longer than a single hash-operation with given input. I'm not talking about all occurrences where the SHA-1-class is used (e.g. during MAC negotiations).
